src/node_report_module.cc

```diff
@@ -0,0 +1,506 @@
+#include "node_internals.h"
+
+#include <sstream>
+#include <v8.h>
+
+using v8::FunctionCallbackInfo;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::NewStringType;
+using v8::StackTrace;
+using v8::String;
+using v8::V8;
+using v8::Value;
+
+namespace nodereport {
+
+// Internal/static function declarations
+static void OnFatalError(const char* location, const char* message);
+bool OnUncaughtException(Isolate* isolate);
+#ifdef _WIN32
+static void PrintStackFromStackTrace(Isolate* isolate, FILE* fp);
+#else  // signal trigger functions for Unix platforms and OSX
+static void SignalDumpAsyncCallback(uv_async_t* handle);
+inline void* ReportSignalThreadMain(void* unused);
+static int StartWatchdogThread(void* (*thread_main)(void* unused));
+static void RegisterSignalHandler(int signo, void (*handler)(int),
+                                  struct sigaction* saved_sa);
+static void RestoreSignalHandler(int signo, struct sigaction* saved_sa);
+static void SignalDump(int signo);
+static void SetupSignalHandler();
+#endif
+
+// Default node-report option settings
+static unsigned int nodereport_events = NR_APICALL;
+static unsigned int nodereport_verbose = 0;
+#ifdef _WIN32  // signal trigger not supported on Windows
+static unsigned int nodereport_signal = 0;
+#else  // trigger signal supported on Unix platforms and OSX
+static unsigned int nodereport_signal = SIGUSR2;  // default signal is SIGUSR2
+static int report_signal = 0;  // atomic for signal handling in progress
+static uv_sem_t report_semaphore;  // semaphore for hand-off to watchdog
+static uv_async_t nodereport_trigger_async;  // async handle for event loop
+static uv_mutex_t node_isolate_mutex;  // mutex for watchdog thread
+static struct sigaction saved_sa;  // saved signal action
+#endif
+
+// State variables for v8 hooks and signal initialisation
+static bool exception_hook_initialised = false;
+static bool error_hook_initialised = false;
+static bool signal_thread_initialised = false;
+
+static Isolate* node_isolate;
+
+/*******************************************************************************
+ * External JavaScript API for triggering a report
+ *
+ ******************************************************************************/
+void TriggerReport(const FunctionCallbackInfo<Value>& info) {
+  Isolate* isolate = info.GetIsolate();
+  HandleScope scope(isolate);
+  char filename[NR_MAXNAME + 1] = "";
+  MaybeLocal<Value> error;
+  int err_index = 0;
+
+  if (info[0]->IsString()) {
+    // Filename parameter supplied
+    String::Utf8Value filename_parameter(isolate, info[0]);
+    if (filename_parameter.length() < NR_MAXNAME) {
+      snprintf(filename, sizeof(filename), "%s", *filename_parameter);
+    } else {
+      isolate->ThrowException(String::NewFromUtf8(
+                              isolate,
+                              "node-report: filename parameter is too long",
+                              NewStringType::kNormal).ToLocalChecked());
```

Comment:
You are throwing a `string` here, not an error object – that doesn’t seem intended?

Ideally, this would use `node_errors.h`.

Reply:
done. added a new error code (ERR_RESOURCE_INITIALIZATION) in `node_errors.h` to reuse for a few scenarios, and thrown as objects.

Reply:
further work on this based on review comments transformed the variable that held the filename into a C++ string, so the filename length limitation went away; along with the need of an error and an error message.